### DISCUSSION TITLE
Can I override few of the binding configuration settings after it has been setup through the config file in the CoreWCF library??

### DISCUSSION THREAD

**Discussion by ubendranr** [CoreWCF/CoreWCF#1319] [Q&A]
Hi
I have recently migrated my WCF services from the .Net Framework (v4.8) to .Net 6.0 through COREWCF Library. However, in my Core WCF project I am loading all the server side config details using COREWCF library through the config file as shown below.
                     services.AddServiceModelServices()
                    .AddServiceModelConfigurationManagerFile(configuration Filename)
                    .AddServiceModelMetadata()
                    .Add Singleton<IServiceBehavior, UseRequestHeadersForMetadataAddressBehavior>();
However, due to some requirement we had to change some of these configuration details in the code, from the one applied through the config file in the COREWCF library so I have played around in the code to see if that is possible at all? After trying with few ideas I have decided to use middleware in my .NET app to allow the basic http binding's security mode to be overridden through the code which has been initially set through the wcf.config file in the CoreWCF library, and here is the code snippet:
	var serviceBasicHttpBindingBehavior = app.ApplicationServices.GetRequiredService<BasicHttpBinding>();
	//If the users have enable the secure transport mode then set the basic security mode to be Transport. 
              serviceBasicHttpBindingBehavior.Security.Mode = secureModeActive ?
                                                                BasicHttpSecurityMode.Transport : BasicHttpSecurityMode.None;
But the changes I am trying to apply to the code is either too early or too late in the pipeline to these changes to take effect? Would it possible to override the settings after the config settings have been applied through the config file in the COREWCF library but before the host start to run so by the time service up and running my new settings will take affect? As you could see from the above snippet we know when we're loading the file but not sure when COREWCF library is applying those settings so it would be nice to know when/where in the code I am able to apply my settings changes so it will take effect before the service start.

Thank in advance!

thanks,
Raj

**Comment by birojnayak** [comment]
I don't think there is a way AFAIK. As soon as you pass the file(config file), we store whole section internally [here](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/ConfigurationManagerServiceModelOptions.cs#L45). Once your hosts start, the configure method is called to act on the binding information present in the section [variable](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/ConfigurationManagerServiceModelOptions.cs#L18) declared inside the `ConfigurationManagerServiceModelOptions` class. There is no public API in between to capture information to modify.

Tagging @mconnew too..

**Reply by mconnew** [reply]
You do get a chance to make changes, but not between loading the config and passing it to CoreWCF. Well, I say that, I think we did expose an abstraction to do that in theory, but that would be messy and harder to work with. Instead use the `IServiceBuilder.ConfigureServiceHostBase<TService>` extension method. You pass it a delegate which takes a `ServiceHostBase` instance. From there, you can iterate over `ServiceHostBase.Description.Endpoints` to examine each of the `ServiceEndpoint` instances. The `ServiceEndpoint.Binding` property has a getter and a setter so you can get the binding and modify it, or replace it altogether. This extensibility point gets called before the equivalent of calling `ServiceHost.Open()` so the entire state and config is able to be modified. So something like this:
```c#
serviceBuilder.ConfigureServiceHostBase<MyService>(serviceHostBase =>
{
    foreach(var ep in serviceHostBase.Description.Endpoints)
    {
        if (ep.Contract.ContractType == typeof(IMyContract))
        {
            var binding = (ep.Binding as BasicHttpBinding) ?? throw new Exception("Not a basic http binding");
            binding.Security.Mode = secureModeActive ? BasicHttpSecurityMode.Transport :
                                                       BasicHttpSecurityMode.None;
        }
    }
});
```

**Reply by ubendranr** [reply]
Thanks for the Suggestion Matt, I will come back to you with feedback/any question once I tried this.

**Comment by ubendranr** [comment]
Hey Matt,

As you have outlined above I have tried to implement my changes at this extension point so this extensibility point is called before the host is start unfortunately we didn't get the desired result as binding settings we're overridden by the config file settings. here is the workflow:

protected override IWebHostBuilder CreateHostBuilder(string[] args, string urls)
{
	  return WebHost.CreateDefaultBuilder()
					.ConfigureHostForSOAP(this._addressConfig, this._unity)
					.ConfigureServices(this.ConfigureServices)
					.Configure(this.Configure);
}
protected override void ConfigureServices(WebHostBuilderContext context, IServiceCollection services)
{
	services.EnableSOAPSupport(this._unity, context.Configuration);
}

protected override void Configure(WebHostBuilderContext context, IApplicationBuilder app)
{
	app.ConfigureSOAPBehaviours(context.Configuration);
   
	app.UseServiceModel(builder =>
	{
		builder.ConfigureAllServiceHostBase(serviceHostBase =>
		{
			foreach (var ep in serviceHostBase.Description.Endpoints)
			{
				//if (ep.Contract.ContractType == typeof(IEDSService))
				//{

				//var binding = (ep.Binding as BasicHttpBinding) ?? throw new Exception("Not a basic http binding");

				var binding = (ep.Binding as BasicHttpBinding);
				if (binding != null)
					binding.Security.Mode = _addressConfig.EnableSecureTransportModes ? BasicHttpSecurityMode.Transport :
																BasicHttpSecurityMode.None;
				//}
			}
		});
	});
}

As you can see I have introduced my changes last in the workflow (Configure method) however, these changes are still being overridden by config file settings for which we load the file using .AddServiceModelConfigurationManagerFile in the line here in the above method: (is this extension method this config gets loaded but we're not sure when that is being applied)
app.ConfigureSOAPBehaviours(context.Configuration)

Is there anyway we could influence the behaviour here so our binding changes take effect just before the host start?

**Reply by mconnew** [reply]
I expected that to work, I'm surprised it didn't. Another option is to use an IServiceBehavior and modify the binding in the Validate method.

**Reply by ubendranr** [reply]
Hi Matt, Thanks for the feedback again however, would you be able to provide me sample code snippet for the above suggestion please?

**Reply by mconnew** [reply]
Where you configure your services, add this:
```c#
    services.AddSingleton<IServiceBehavior, ModifyBindingServiceBehavior>();
```
The behavior would look something like this:
```c#
    public class ModifyBindingServiceBehavior : IServiceBehavior
    {
        public void AddBindingParameters(ServiceDescription serviceDescription, ServiceHostBase serviceHostBase, Collection<ServiceEndpoint> endpoints, BindingParameterCollection bindingParameters) { }
        public void ApplyDispatchBehavior(ServiceDescription serviceDescription, ServiceHostBase serviceHostBase) { }

        public void Validate(ServiceDescription serviceDescription, ServiceHostBase serviceHostBase)
        {
            foreach (var endpoint in serviceHostBase.Description.Endpoints)
            {
                if (endpoint.Contract.ContractType == typeof(IService))
                {
                    var binding = new CustomBinding(endpoint.Binding);
                    // Make changes
                    endpoint.Binding = binding;
                }
            }
        }
    }
```
As it's coming from DI, you can inject into the constructor anything you need to get the addressConfig instance.

**Reply by ngvtien** [reply]
Yeah it didn't take effect because `builder.Services` is still being built which is empty at the time `ConfigureAllServiceHostBase` is called.
